pkg/extract/oracle/oracle.go

```diff
@@ -41,24 +72,482 @@ func Extract(args string, opts ...Option) error {
+				Source: "linux.oracle.com/security",
+				Vendor: &ovalDef.Metadata.Advisory.Severity}},
+			Published: utiltime.Parse([]string{"2006-01-02T15:04:05Z"}, ovalDef.Metadata.Advisory.Issued.Date),
+		}},
+		DataSource: sourceTypes.Oracle,
+	}
+
+	refURLs := map[string]struct{}{}             // URL -> struct{}
+	vulnURLs := map[string]map[string]struct{}{} // vuln ID -> URL -> struct{}
+
+	for _, r := range ovalDef.Metadata.Reference {
+		refURLs[r.RefURL] = struct{}{}
+		if r.Source == "CVE" {
+			if vulnURLs[r.RefID] == nil {
+				vulnURLs[r.RefID] = map[string]struct{}{}
+			}
+			vulnURLs[r.RefID][r.RefURL] = struct{}{}
+		}
+	}
+	for _, c := range ovalDef.Metadata.Advisory.Cve {
+		vulnURLs[c.Text][c.Href] = struct{}{}
+	}
+
+	data.Vulnerabilities = func() []vulnerabilityTypes.Vulnerability {
+		vs := make([]vulnerabilityTypes.Vulnerability, 0, len(vulnURLs))
+		for vulnID, urls := range vulnURLs {
+			vs = append(vs, vulnerabilityTypes.Vulnerability{
+				ID: vulnID,
+				References: func() []referenceTypes.Reference {
+					refs := make([]referenceTypes.Reference, 0, len(urls))
+					for url := range urls {
+						refs = append(refs, referenceTypes.Reference{
+							Source: "linux.oracle.com/security",
+							URL:    url,
+						})
+					}
+					return refs
+				}(),
+			})
+		}
+		return vs
+	}()
+
+	data.Advisories[0].References = func() []referenceTypes.Reference {
+		refs := make([]referenceTypes.Reference, 0, len(refURLs))
+		for url := range refURLs {
+			refs = append(refs, referenceTypes.Reference{
+				Source: "linux.oracle.com/security",
+				URL:    url,
+			})
+		}
+		return refs
+	}()
+
+	allPkgs, err := collectPackages(ovalDef.Criteria, tos)
+	if err != nil {
+		return dataTypes.Data{}, errors.Wrapf(err, "collectPackages, definition: %s", ovalDef.ID)
+	}
+
+	data.Detection = func() []detectionTypes.Detection {
+		ds := make([]detectionTypes.Detection, 0, len(allPkgs))
+		for major, pkgs := range allPkgs {
+			ds = append(ds, detectionTypes.Detection{
+				Ecosystem: detectionTypes.Ecosystem(fmt.Sprintf("%s:%s", detectionTypes.EcosystemTypeOracle, major)),
+				Criteria: criteriaTypes.Criteria{
+					Operator: criteriaTypes.CriteriaOperatorTypeOR,
+					Criterions: func() []criterionTypes.Criterion {
+						criterions := make([]criterionTypes.Criterion, 0, len(pkgs))
+						for pkg, archs := range pkgs {
+							criterions = append(criterions, criterionTypes.Criterion{
+								Vulnerable: true,
+								Package: packageTypes.Package{
+									Name:          pkg.name,
+									Architectures: archs,
+								},
+								Affected: &affectedTypes.Affected{
+									Type:  rangeTypes.RangeTypeRPM,
+									Range: []rangeTypes.Range{{LessThan: pkg.fixedVersion}},
+									Fixed: []string{pkg.fixedVersion},
+								},
+							})
+						}
+						return criterions
+					}(),
+				},
+			})
+		}
+		return ds
+	}()
+
+	return data, nil
+}
+
+type ovalPackage struct {
+	major           string
+	ovalName        string
```

Comment:
If `ovalName` is to handle package name, another variable name would be better.

Reply:
Changed to "name" simply, thanks! 3f4ed82